### PR TITLE
adding resource `azuredevops_environment_resource_kubernetes`

### LINKED ISSUE #584: Add ability to add a resource under an Environment - Kubernetes / Virtual Machine
Hi,

I am not able to find the functionality, maybe I have it wrong, but is there a way to create a resource under an environment?  ie: either add a kubernetes namespace or a virtual machine.

We are presently stuck at creating a functioning environment because this is missing (or we are not able to find out how to do it).

The missing piece is supposed to under-the-hood call MS azure devops REST API:
https://docs.microsoft.com/en-us/rest/api/azure/devops/distributedtask/kubernetes/add?view=azure-devops-rest-6.0

POST https://dev.azure.com/{organization}/{project}/_apis/distributedtask/environments/{environmentId}/providers/kubernetes?api-version=6.0-preview.1

We are using the pulumi devops provider which derives from the terraform provider.

Thanks for your time,

### LINKED ISSUE #807: azuredevops_serviceendpoint_kubernetes - Add AKS to Azure DevOps Environment
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and Azure DevOps Provider) Version

Terraform v1.5.2
on linux_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.39.0
+ provider registry.terraform.io/hashicorp/azurerm v3.58.0
+ provider registry.terraform.io/hashicorp/null v3.2.1
+ provider registry.terraform.io/hashicorp/time v0.9.1
+ provider registry.terraform.io/microsoft/azuredevops v0.6.0

### Affected Resource(s)

+ azuredevops_serviceendpoint_kubernetes

### Terraform Configuration Files

Provider.tf

```hcl
terraform {
  required_providers {
    azurerm   = {
      source  = "hashicorp/azurerm"
      version = "=3.58.0"
    }
    null = {
      source  = "hashicorp/null"
      version = "=3.2.1"
    }
    azuread = {
      source = "hashicorp/azuread"
      version = "2.39.0"
    }
    time = {
      source = "hashicorp/time"
      version = "0.9.1"
    }
    azuredevops = {
      source = "microsoft/azuredevops"
      version = "0.6.0"
    }
  }
}

provider "azurerm" {
  features {
    key_vault {
      purge_soft_delete_on_destroy = true
      recover_soft_deleted_key_vaults = true
    }
  }
}

provider "azuredevops" {
  org_service_url       = "https://dev.azure.com/xxxxxxx"
  personal_access_token = data.azurerm_key_vault_secret.terra_kva_secret_personal_access_token.value
}
```

AKS Environment
```hcl
data "azuredevops_project" "devops_projet" {
  project_id = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
}

resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_service_endpoint" {
  project_id            = data.azuredevops_project.devops_projet.id
  service_endpoint_name = "aks01${local.environnement}"
  apiserver_url         = "https://${azurerm_kubernetes_cluster.terra_aks.fqdn}"
  authorization_type    = "AzureSubscription"

  azure_subscription {
    subscription_id   = data.azurerm_key_vault_secret.terra_kva_secret_subsid.value
    subscription_name = "${local.environnement}"
    tenant_id         = data.azurerm_key_vault_secret.terra_kva_secret_tenantid.value
    resourcegroup_id  = azurerm_resource_group.terra_rgo.id
    cluster_name      = azurerm_kubernetes_cluster.terra_aks.name
  }
}

resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_kubeconfig" {
  project_id            = data.azuredevops_project.devops_projet.id
  service_endpoint_name = "aks01${local.environnement}"
  apiserver_url         = "https://${azurerm_kubernetes_cluster.terra_aks.fqdn}"
  authorization_type    = "Kubeconfig"

  kubeconfig {
    kube_config            = azurerm_kubernetes_cluster.terra_aks.kube_admin_config_raw
  }
}

resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_service_endpoint_service_account" {
  project_id            = data.azuredevops_project.devops_projet.id
  service_endpoint_name = "aks01${local.environnement}"
  apiserver_url         = "https://${azurerm_kubernetes_cluster.terra_aks.fqdn}"
  authorization_type    = "ServiceAccount"

  service_account {
    token   = yamldecode(azurerm_kubernetes_cluster.terra_aks.kube_config_raw).users.0.user.token
    ca_cert = azurerm_kubernetes_cluster.terra_aks.kube_config.0.cluster_ca_certificate
  }
}
```

### Debug Output

github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xf92380?, {0xf92380?, 0xc00039def0?}, 0xd?, {0xca4380?, 0xc00034a780?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000345420, {0xf92380, 0xc00039def0}, 0xc000582820, 0xc000409400, {0xca4380, 0xc00034a780})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/resource.go:837 +0xa7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0002fea50, {0xf92380?, 0xc00039ddd0?}, 0xc000141720)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/grpc_provider.go:1021 +0xe3c
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0002d8fa0, {0xf92380?, 0xc00039d5c0?}, 0xc00041d730)
	github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xdfe6a0?, 0xc0002d8fa0}, {0xf92380, 0xc00039d5c0}, 0xc00041d6c0, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001e21e0, {0xf95040, 0xc000003860}, 0xc000464d80, 0xc00035d3e0, 0x14b7760, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1295 +0xb0b
google.golang.org/grpc.(*Server).handleStream(0xc0001e21e0, {0xf95040, 0xc000003860}, 0xc000464d80, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1636 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.48.0/server.go:930 +0x28a

Error: The terraform-provider-azuredevops_v0.6.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

##[error]Bash exited with code '1'.

### Expected Behavior

The code needs to add an AKS in an Azure DevOps environment (Pipelines > Environments). The problem is with the endpoint service for the Azure service account in AKS.

### Actual Behavior

The Terraform plan output is : 

```hcl
Terraform will perform the following actions:

  # azuredevops_serviceendpoint_kubernetes.terra_aks_kubeconfig will be created
  + resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_kubeconfig" {
      + apiserver_url         = "[https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io"](https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io%22/)
      + authorization         = (known after apply)
      + authorization_type    = "Kubeconfig"
      + description           = "Managed by Terraform"
      + id                    = (known after apply)
      + project_id            = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
      + service_endpoint_name = "aks01int01"

      + kubeconfig {
          + accept_untrusted_certs = true
          + cluster_context        = (known after apply)
        }
    }

  # azuredevops_serviceendpoint_kubernetes.terra_aks_service_endpoint will be created
  + resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_service_endpoint" {
      + apiserver_url         = "[https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io"](https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io%22/)
      + authorization         = (known after apply)
      + authorization_type    = "AzureSubscription"
      + description           = "Managed by Terraform"
      + id                    = (known after apply)
      + project_id            = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
      + service_endpoint_name = "aks01int01"

      + azure_subscription {
          # At least one attribute in this block is (or was) sensitive,
          # so its contents will not be displayed.
        }
    }

  # azuredevops_serviceendpoint_kubernetes.terra_aks_service_endpoint_service_account will be created
  + resource "azuredevops_serviceendpoint_kubernetes" "terra_aks_service_endpoint_service_account" {
      + apiserver_url         = "[https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io"](https://aks01rgo01int01-xxxxxxxx.hcp.francecentral.azmk8s.io%22/)
      + authorization         = (known after apply)
      + authorization_type    = "ServiceAccount"
      + description           = "Managed by Terraform"
      + id                    = (known after apply)
```

### Steps to Reproduce

On Azure, you need : 
+ A resource group
+ An Azure Kubernetes Services
+ An Azure DevOps organization and project already configured
+ A personal access token from Azure DevOps which can be available to add AKS on Azure DevOps

1. `terraform plan`
2. `terraform apply`

### PR DESCRIPTION
**Pull request by rdalbuquerque**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?
Added the ability to manage Kubernetes environment resources.

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This is the continuation of PR #553 , now with default `azure-devops-go-api` `taskagent` client

### CODE REVIEW

**Comment by xuzhang3** [comment]
@rdalbuquerque  can you help resolve the file conflict

**Comment by rdalbuquerque** [comment]
> @rdalbuquerque can you help resolve the file conflict

Done!

**Comment by rdalbuquerque** [comment]
Hi @xuzhang3 ,
Anything else needed on this PR?

File: azuredevops/internal/service/taskagent/resource_kubernetes_resource.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
+package taskagent
```
Comment (xuzhang3) [inline azuredevops/internal/service/taskagent/resource_kubernetes_resource.go]: rename `resource_kubernetes_resource.go` to `resource_environment_resource_kubernetes.go`

File: azuredevops/internal/service/taskagent/resource_kubernetes_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
+//go:build (all || resource_environment_resource_kubernetes) && !exclude_resource_environment_resource_kubernetes
```
Comment (xuzhang3) [inline azuredevops/internal/service/taskagent/resource_kubernetes_resource_test.go]: rename `resource_kubernetes_resource_test.go` to `resource_environment_resource_kubernetes_test.go`

File: azuredevops/internal/service/taskagent/resource_kubernetes_resource.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			ServiceEndpointId: resource.ServiceEndpointId,
+		},
+		Project:       converter.String(project.Id.String()),
+		EnvironmentId: resource.EnvironmentReference.Id,
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating Kubernetes resource in Azure DevOps: %+v", err)
+	}
+
+	flattenKubernetesResource(d, project, createdResource)
```
Comment (xuzhang3) [inline azuredevops/internal/service/taskagent/resource_kubernetes_resource.go]: Create function should end with a call to the read function

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Review by xuzhang3 — APPROVED** [review]
```log
=== RUN   TestAccEnvironmentKubernetes_CreateAndUpdate
=== PAUSE TestAccEnvironmentKubernetes_CreateAndUpdate
=== CONT  TestAccEnvironmentKubernetes_CreateAndUpdate
--- PASS: TestAccEnvironmentKubernetes_CreateAndUpdate (145.46s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        165.251s

```

**Pull request merged**